File: ui/src/lib/backend/language-servers.ts
/**
 * What can answer a question about meaning in the folder he is working in.
 *
 * The semantic tools — references, definition, rename — need a language server, and most
 * machines have none. That is ordinary rather than broken, and it is reported as status
 * with a remedy rather than as an error.
 *
 * Installing from here involves no permission prompt, and that is deliberate: the gate
 * exists to stop *Kith* changing your machine without asking, and a person clicking a
 * button in their own settings window is the asking. His own route through the chat still
 * goes through the gate, because there the actor is him.
 */

/** One *server* the folder needs, and the languages it would answer for. */
export type LanguageServerState = {
  /** The family that answers — "python", "typescript". One server per family. */
  family: string;
  /** The languages in this folder that this one server covers, most files first.
   *
   * **A row is a server, not a language**, and it used to be the other way round. `tsx`,
   * `javascript` and `typescript` are three separate answers from the outline and one
   * `typescript-language-server` serves all three — so a TypeScript project showed three rows
   * each offering the same 32 MB download, and installing from any of them satisfied all three
   * at once, which read as the button having done nothing. */
  languages: string[];
  /** How many files this server would cover, across all of them. */
  files: number;
  /** Whether something can already answer semantic questions about it. */
  served: boolean;
  /** Whether Kith can fetch one. False for Go, Rust, Ruby and C, which come from their own
   *  package managers — `manual` carries the command for those. */
  installable: boolean;
  /** Roughly what lands on disk, in words. Empty when we do not install it. */
  size: string;
  /** What to run by hand. Always present, and the only option when `installable` is false. */
  manual: string;
};

export type LanguageServers = {
  /** The folder inspected. Shown because "this needs pyright" is only useful next to which
   *  project — and it follows the work rather than sitting at the workspace root. */
  root: string;
  languages: LanguageServerState[];
};

export async function fetchLanguageServers(): Promise<LanguageServers> {
  const response = await fetch("/api/language-servers");
  if (!response.ok) throw new Error(`could not read language support (${response.status})`);
  return (await response.json()) as LanguageServers;
}

/** Install the one for this language. Available on his next call, with no restart. */
export async function installLanguageServer(family: string): Promise<void> {
  const response = await fetch("/api/language-servers/install", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ family }),
  });
  const body = (await response.json()) as { error?: string };
  if (!response.ok) throw new Error(body.error ?? `installing failed (${response.status})`);
}

File: server/kith/api/routes/language_servers.py
"""What can answer a question about meaning here, and getting one that can.

The semantic tools — who calls this, where is it defined, what breaks if I rename it — need a
language server, and most machines have none. That is an ordinary condition rather than a
broken one, but until now the only way out of it was through a conversation: he noticed, he
offered, you approved, he installed. That works and costs a turn.

This is the other path. The settings page can ask what this folder is written in, what can
serve it, and install the one thing missing — for nothing, on any turn, without involving him
at all.

**No permission prompt on this route, deliberately.** The gate exists to stop *Kith* doing
things to your machine without asking. A person clicking Install in their own settings window
is the asking; putting a second dialog in front of that would be the app confirming that you
meant to press the button you pressed. His route through `install_language_support` still goes
through the gate, because there the actor is him.

It reports the folder it looked at. "This project needs pyright" is only useful next to which
project — and `base_dir()` follows the work rather than sitting at the workspace root, which is
the difference between an answer about the code you are looking at and an answer about an empty
folder.
"""

from __future__ import annotations

from flask import jsonify, request

from kith.api.blueprint import api
from kith.engine.code.lsp import install
from kith.engine.code.lsp.manager import CANDIDATES, manager


def _served(root):
    def check(family: str) -> bool:
        try:
            return manager.find_binary(family, root) is not None
        except Exception:
            return False

    return check


@api.get("/language-servers")
@api.doc(
    summary="Which languages this folder uses, and what can serve each one",
    description=(
        "Reports the folder inspected and one entry per language found in it: how many "
        "files, whether a server is already available, whether Kith can install one, and "
        "roughly how large it is. Languages whose servers come from their own package "
        "manager (Go, Rust, Ruby, C) are reported as not installable, with the command."
    ),
)
def language_servers_state():
    from kith.infra import workspace as sandbox

    root = sandbox.base_dir()
    served = _served(root)
    counted = install.languages_in(root)

    # One row per *server*, not per language.
    #
    # `languages_in` counts by language — `tsx`, `javascript` and `typescript` are three
    # different answers — and this reported each of them separately. So a project written in
    # TypeScript showed three rows, each offering a 32 MB download, for the one
    # `typescript-language-server` that serves all three. Installing from any of them satisfied
    # all three at once, which read as the button doing nothing.
    #
    # Grouped by the family that actually answers, with the languages it covers named so the
    # row still says why it is there.
    grouped: dict[str, dict] = {}
    for language, files in sorted(counted.items(), key=lambda pair: -pair[1]):
        if files < install.MIN_FILES:
            continue  # one stray file is not what this project is written in
        family = manager.family_of(language)
        entry = grouped.get(family)
        if entry is None:
            candidates = CANDIDATES.get(family, ())
            entry = grouped[family] = {
                "family": family,
                "languages": [],
                "files": 0,
                "served": served(family),
                # Asked of the *language* rather than the family: `install.PACKAGES` is keyed by
                # language, and a family name is not always one of its keys.
                "installable": bool(install.installable(language)),
                "size": install.download_size(language),
                # What to run by hand for the ones we do not install. The first candidate is
                # the preferred one, and its own table already carries the command.
                "manual": candidates[0].install if candidates else "",
            }
        entry["languages"].append(language)
        entry["files"] += files

    languages = sorted(grouped.values(), key=lambda one: -one["files"])
    return jsonify({"root": str(root), "languages": languages})


@api.post("/language-servers/install")
@api.doc(
    summary="Install the language server for one language",
    description=(
        "Body {family: 'python'}. Installs into Kith's own folder, never globally and never "
        "into the application bundle, and only from a closed table of packages — the family "
        "name selects a command, it never becomes one. Available on the next call with no "
        "restart."
    ),
)
def language_servers_install():
    body = request.get_json(silent=True) or {}
    family = str(body.get("family") or "").strip().lower()
    if not family:
        return jsonify({"error": "family is required"}), 400
    if not install.installable(family):
        return jsonify({"error": f"there is no server I install for {family}"}), 400

    worked, said = install.run(family)
    if not worked:
        return jsonify({"error": said}), 400
    return jsonify({"ok": True, "family": family, "where": said})
